File: desktop/src/local-ai-service/src/services/activity_classifier_ml.py
"""
ML-based Activity Classifier using transformers
Tier-based: HIGH_END and PREMIUM use ML, LOW_END/MID_RANGE use rule-based
"""
from typing import Dict, Optional, List
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
from src.services.model_manager import get_model_manager
from src.config import settings
from src.utils.logger import setup_logger

logger = setup_logger()

# Global model instances (lazy loading)
_classifier_model = None
_classifier_tokenizer = None
_device = None

# Activity type labels
ACTIVITY_TYPES = [
    'coding',
    'reading',
    'watching',
    'gaming',
    'shopping',
    'social',
    'learning',
    'entertainment',
    'work',
    'other',
]

def get_device():
    """Get device (CPU or CUDA) with fallback to CPU"""
    global _device
    if _device is None:
        if torch.cuda.is_available():
            try:
                test_tensor = torch.zeros(1).cuda()
                _device = 'cuda'
                logger.info("Using CUDA for activity classifier")
            except Exception as e:
                logger.warning(f"CUDA available but not usable: {e}. Falling back to CPU")
                _device = 'cpu'
        else:
            _device = 'cpu'
            logger.info("Using CPU for activity classifier (CUDA not available)")
    return _device

def load_classifier_model():
    """Load classifier model based on system tier"""
    global _classifier_model, _classifier_tokenizer
    
    if _classifier_model is not None:
        return _classifier_model, _classifier_tokenizer
    
    model_manager = get_model_manager()
    tier = model_manager.get_recommended_tier() if not settings.MODEL_TIER else settings.MODEL_TIER
    
    # Only use ML classifier for HIGH_END and PREMIUM tiers
    if tier not in ['HIGH_END', 'PREMIUM']:
        logger.info(f"Tier {tier} uses rule-based classifier, not loading ML model")
        return None, None
    
    model_name = "distilbert-base-uncased"
    device = get_device()
    
    try:
        logger.info(f"Loading activity classifier model: {model_name} on {device}")
        
        # Load tokenizer
        _classifier_tokenizer = AutoTokenizer.from_pretrained(model_name)
        
        # For now, we'll use a zero-shot classification approach
        # In production, you'd fine-tune this model on labeled activity data
        # Using a zero-shot classifier as a starting point
        from transformers import pipeline
        
        # Create zero-shot classification pipeline
        classifier = pipeline(
            "zero-shot-classification",
            model=model_name,
            device=0 if device == 'cuda' else -1,
        )
        
        _classifier_model = classifier
        logger.info(f"Activity classifier model loaded: {model_name}")
        
        return _classifier_model, _classifier_tokenizer
    except Exception as e:
        logger.error(f"Error loading classifier model: {e}")
        return None, None

def should_use_ml_classifier() -> bool:
    """Determine if ML classifier should be used based on tier"""
    model_manager = get_model_manager()
    tier = model_manager.get_recommended_tier() if not settings.MODEL_TIER else settings.MODEL_TIER
    return tier in ['HIGH_END', 'PREMIUM']

async def classify_activity_ml(
    app_name: str,
    window_title: str,
    url: Optional[str] = None,
    content_snippet: Optional[str] = None,
) -> Dict:
    """
    Classify activity using ML model
    
    Args:
        app_name: Application name
        window_title: Window title
        url: URL (if browser activity)
        content_snippet: Content snippet (optional)
    
    Returns:
        Dict with 'activity_type', 'confidence', 'metadata', 'reason'
    """
    try:
        # Check if we should use ML classifier
        if not should_use_ml_classifier():
            return {
                'activity_type': 'other',
                'confidence': 0.5,
                'metadata': {},
                'reason': 'Tier does not support ML classification, use rule-based',
            }
        
        # Load model if not loaded
        classifier, tokenizer = load_classifier_model()
        
        if classifier is None:
            logger.warning("ML classifier not available, falling back to rule-based")
            return {
                'activity_type': 'other',
                'confidence': 0.5,
                'metadata': {},
                'reason': 'ML model not loaded, use rule-based classifier',
            }
        
        # Prepare input text
        input_text = f"{app_name} {window_title}"
        if url:
            input_text += f" {url}"
        if content_snippet:
            # Limit content snippet to first 200 chars
            input_text += f" {content_snippet[:200]}"
        
        # Classify using zero-shot classification
        result = classifier(input_text, ACTIVITY_TYPES)
        
        # Extract top prediction
        if result and 'labels' in result and 'scores' in result:
            top_label = result['labels'][0]
            top_score = result['scores'][0]
            
            # Map to our activity types
            activity_type = map_to_activity_type(top_label)
            
            return {
                'activity_type': activity_type,
                'confidence': float(top_score),
                'metadata': {
                    'all_predictions': {
                        label: float(score)
                        for label, score in zip(result['labels'], result['scores'])
                    },
                    'model': 'distilbert-base-uncased',
                    'method': 'zero-shot-classification',
                },
                'reason': f'ML classification: {top_label} (confidence: {top_score:.2f})',
            }
        else:
            return {
                'activity_type': 'other',
                'confidence': 0.5,
                'metadata': {},
                'reason': 'ML classification returned unexpected format',
            }
    
    except Exception as e:
        logger.error(f"Error in ML activity classification: {e}")
        return {
            'activity_type': 'other',
            'confidence': 0.5,
            'metadata': {'error': str(e)},
            'reason': f'ML classification error: {str(e)}',
        }

def map_to_activity_type(label: str) -> str:
    """Map classifier label to our activity type"""
    # Direct mapping
    if label in ACTIVITY_TYPES:
        return label
    
    # Fallback mappings
    mappings = {
        'code': 'coding',
        'programming': 'coding',
        'editor': 'coding',
        'read': 'reading',
        'book': 'reading',
        'pdf': 'reading',
        'video': 'watching',
        'youtube': 'watching',
        'stream': 'watching',
        'game': 'gaming',
        'play': 'gaming',
        'shop': 'shopping',
        'buy': 'shopping',
        'ecommerce': 'shopping',
        'social': 'social',
        'media': 'social',
        'learn': 'learning',
        'study': 'learning',
        'tutorial': 'learning',
        'entertain': 'entertainment',
        'movie': 'entertainment',
        'music': 'entertainment',
        'work': 'work',
        'office': 'work',
    }
    
    label_lower = label.lower()
    for key, value in mappings.items():
        if key in label_lower:
            return value
    
    return 'other'

async def batch_classify_activities(activities: List[Dict]) -> List[Dict]:
    """Classify multiple activities in batch"""
    results = []
    
    for activity in activities:
        result = await classify_activity_ml(
            app_name=activity.get('app_name', ''),
            window_title=activity.get('window_title', ''),
            url=activity.get('url'),
            content_snippet=activity.get('content_snippet'),
        )
        results.append(result)
    
    return results

File: desktop/src/main/services/activity/activity-classifier-ml.js
import axios from 'axios';
import logger from '../utils/logger.js';
import { getAppConfig } from '../utils/config-manager.js';
import { classifyActivity as ruleBasedClassify } from '../filters/activity-classifier.js';

// Cache for recent classifications
const classificationCache = new Map();
const CACHE_SIZE = 100;
const CACHE_TTL = 5 * 60 * 1000; // 5 minutes

// AI service base URL
function getAIServiceURL() {
    const config = getAppConfig();
    return config.aiServiceURL || 'http://127.0.0.1:8000';
}

/**
 * Check if ML classifier is available
 * @returns {Promise<boolean>}
 */
async function isMLClassifierAvailable() {
    try {
        const response = await axios.get(`${getAIServiceURL()}/api/v1/classifier/status`, {
            timeout: 2000,
        });
        return response.data?.ml_available === true;
    } catch (error) {
        logger.debug('ML classifier not available, using rule-based:', error.message);
        return false;
    }
}

/**
 * Classify activity using ML (with fallback to rule-based)
 * @param {Object} activity - Activity object
 * @returns {Promise<Object>} - Classification result
 */
async function classifyActivityML(activity) {
    try {
        // Check cache first
        const cacheKey = `${activity.app_name}|${activity.window_title}|${activity.url || ''}`;
        const cached = classificationCache.get(cacheKey);

        if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
            logger.debug('Using cached classification');
            return cached.result;
        }

        // Check if ML classifier is available
        const mlAvailable = await isMLClassifierAvailable();

        if (!mlAvailable) {
            // Fallback to rule-based
            logger.debug('ML classifier not available, using rule-based');
            const result = ruleBasedClassify(activity);

            // Cache result
            cacheResult(cacheKey, result);
            return result;
        }

        // Use ML classifier
        try {
            const response = await axios.post(
                `${getAIServiceURL()}/api/v1/classify-activity`,
                {
                    app_name: activity.app_name || '',
                    window_title: activity.window_title || '',
                    url: activity.url || null,
                    content_snippet: activity.content?.substring(0, 200) || null,
                },
                {
                    timeout: 5000, // 5 second timeout
                }
            );

            const result = {
                type: response.data.activity_type || 'other',
                confidence: response.data.confidence || 0.5,
                reason: response.data.reason || 'ML classification',
                metadata: response.data.metadata || {},
            };

            // Cache result
            cacheResult(cacheKey, result);

            logger.debug(`ML classification: ${result.type} (confidence: ${result.confidence})`);
            return result;
        } catch (mlError) {
            // ML service error, fallback to rule-based
            logger.warn('ML classifier error, falling back to rule-based:', mlError.message);
            const result = ruleBasedClassify(activity);
            cacheResult(cacheKey, result);
            return result;
        }
    } catch (error) {
        logger.error('Error in ML classification:', error);
        // Always fallback to rule-based
        return ruleBasedClassify(activity);
    }
}

/**
 * Batch classify activities
 * @param {Array<Object>} activities - Array of activity objects
 * @returns {Promise<Array<Object>>} - Array of classification results
 */
async function batchClassifyActivities(activities) {
    try {
        const mlAvailable = await isMLClassifierAvailable();

        if (!mlAvailable || activities.length === 0) {
            // Fallback to rule-based for each
            return activities.map(activity => ruleBasedClassify(activity));
        }

        try {
            const response = await axios.post(
                `${getAIServiceURL()}/api/v1/batch-classify`,
                {
                    activities: activities.map(activity => ({
                        app_name: activity.app_name || '',
                        window_title: activity.window_title || '',
                        url: activity.url || null,
                        content_snippet: activity.content?.substring(0, 200) || null,
                    })),
                },
                {
                    timeout: 10000, // 10 second timeout for batch
                }
            );

            return response.data.results.map(result => ({
                type: result.activity_type || 'other',
                confidence: result.confidence || 0.5,
                reason: result.reason || 'ML batch classification',
                metadata: result.metadata || {},
            }));
        } catch (mlError) {
            logger.warn('ML batch classification error, falling back to rule-based:', mlError.message);
            return activities.map(activity => ruleBasedClassify(activity));
        }
    } catch (error) {
        logger.error('Error in batch ML classification:', error);
        return activities.map(activity => ruleBasedClassify(activity));
    }
}

/**
 * Cache classification result
 */
function cacheResult(key, result) {
    // Limit cache size
    if (classificationCache.size >= CACHE_SIZE) {
        // Remove oldest entry
        const firstKey = classificationCache.keys().next().value;
        classificationCache.delete(firstKey);
    }

    classificationCache.set(key, {
        result,
        timestamp: Date.now(),
    });
}

/**
 * Clear classification cache
 */
function clearCache() {
    classificationCache.clear();
}

export {
    classifyActivityML,
    batchClassifyActivities,
    isMLClassifierAvailable,
    clearCache,
};